pkg/tcpip/stack/stack.go

```diff
@@ -1161,17 +1168,32 @@ func (s *Stack) FindRoute(id tcpip.NICID, localAddr, remoteAddr tcpip.Address, n
 				continue
 			}
 
+			plen := route.Destination.Mask().Prefix()
+			if plen < bestPrefixLen {
+				continue
+			}
+
 			if id == 0 || id == route.NIC {
 				if addressEndpoint := s.getAddressEP(nic, localAddr, remoteAddr, netProto); addressEndpoint != nil {
 					var gateway tcpip.Address
 					if needRoute {
 						gateway = route.Gateway
 					}
-					r := constructAndValidateRoute(netProto, addressEndpoint, nic /* outgoingNIC */, nic /* outgoingNIC */, gateway, localAddr, remoteAddr, s.handleLocal, multicastLoop)
-					if r == nil {
+
+					if r := constructAndValidateRoute(netProto, addressEndpoint, nic /* outgoingNIC */, nic /* outgoingNIC */, gateway, localAddr, remoteAddr, s.handleLocal, multicastLoop); r == nil {
+						panic(fmt.Sprintf("non-forwarding route validation failed with route table entry = %#v, id = %d, localAddr = %s, remoteAddr = %s", route, id, localAddr, remoteAddr))
+					} else if plen >= bestPrefixLen {
+						bestRoute, bestPrefixLen = r, plen
+					}
+				} else {
+					// There's no local address on the NIC but this route suggests that
```

Comment:
How does this change affect route selection for say a UDP endpoint in the following scenario:

- NIC_A w/ no addresses
- NIC_B w/ an address
- default route through NIC_B
- specific subnet route through NIC_A



Create a UDP socket and attempt to connect to some destination that initially matches the more specific route through NIC_A. What is the socket's local address?

Reply:
I've updated the PR so that it keeps a track of the best local address for interfaces which pass `getAddressEP` and `constructAndValidateRoute`. This should mean that the source address for the next-closest interface with an address should be chosen if possible.

Reply:
We do `bestLocalAddr = localAddr` now but `localAddr` is never updated. Am i missing something?
 
Can we test this scenario?